File: backend/src/main/java/com/technicalescaperoom/backend/service/EventService.java
package com.technicalescaperoom.backend.service;

import com.technicalescaperoom.backend.dto.admin.CreateEventRequest;
import com.technicalescaperoom.backend.dto.admin.EventResponse;
import com.technicalescaperoom.backend.dto.admin.UpdateEventRequest;
import com.technicalescaperoom.backend.entity.Event;
import com.technicalescaperoom.backend.enums.EventStatus;
import com.technicalescaperoom.backend.enums.GameEventType;
import com.technicalescaperoom.backend.exception.ResourceNotFoundException;
import com.technicalescaperoom.backend.repository.EventRepository;
import com.technicalescaperoom.backend.repository.TeamRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.security.SecureRandom;
import java.time.Instant;
import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class EventService {

    private final EventRepository eventRepository;
    private final TeamRepository teamRepository;
    private final AuditService auditService;
    private final SecureRandom random = new SecureRandom();

    @Transactional
    public EventResponse createEvent(CreateEventRequest request) {
        validateTimestamps(request.getStartTime(), request.getEndTime());

        String rawPasskey = (request.getPasskey() != null && !request.getPasskey().isBlank())
                ? request.getPasskey()
                : generateRandomPasskey();

        // Standard passkey hash representation for Phase 3 storage
        String passkeyHash = "HASH_" + Integer.toHexString(rawPasskey.hashCode());

        Event event = Event.builder()
                .name(request.getName().trim())
                .description(request.getDescription())
                .status(EventStatus.DRAFT)
                .passkeyHash(passkeyHash)
                .startTime(request.getStartTime())
                .endTime(request.getEndTime())
                .build();

        Event saved = eventRepository.save(event);
        auditService.logEvent(GameEventType.EVENT_CREATED, saved, null, null,
                "{\"name\":\"" + saved.getName() + "\"}");

        return mapToResponse(saved);
    }

    @Transactional(readOnly = true)
    public EventResponse getEventById(Long eventId) {
        Event event = findEventOrThrow(eventId);
        return mapToResponse(event);
    }

    @Transactional(readOnly = true)
    public List<EventResponse> getAllEvents() {
        return eventRepository.findAll().stream()
                .map(this::mapToResponse)
                .collect(Collectors.toList());
    }

    @Transactional
    public EventResponse updateEvent(Long eventId, UpdateEventRequest request) {
        validateTimestamps(request.getStartTime(), request.getEndTime());

        Event event = findEventOrThrow(eventId);
        event.setName(request.getName().trim());
        event.setDescription(request.getDescription());
        event.setStatus(request.getStatus());
        event.setStartTime(request.getStartTime());
        event.setEndTime(request.getEndTime());

        Event updated = eventRepository.save(event);
        auditService.logEvent(GameEventType.EVENT_UPDATED, updated, null, null,
                "{\"status\":\"" + updated.getStatus() + "\"}");

        return mapToResponse(updated);
    }

    @Transactional
    public EventResponse updateEventStatus(Long eventId, EventStatus status) {
        Event event = findEventOrThrow(eventId);
        event.setStatus(status);

        Event updated = eventRepository.save(event);
        auditService.logEvent(GameEventType.EVENT_STATUS_CHANGED, updated, null, null,
                "{\"newStatus\":\"" + status + "\"}");

        return mapToResponse(updated);
    }

    public Event findEventOrThrow(Long eventId) {
        return eventRepository.findById(eventId)
                .orElseThrow(() -> new ResourceNotFoundException("Event not found with ID: " + eventId));
    }

    private void validateTimestamps(Instant startTime, Instant endTime) {
        if (startTime != null && endTime != null && endTime.isBefore(startTime)) {
            throw new IllegalArgumentException("Event end time cannot be before start time");
        }
    }

    private String generateRandomPasskey() {
        int number = 100000 + random.nextInt(900000);
        return String.valueOf(number);
    }

    private EventResponse mapToResponse(Event event) {
        long teamCount = teamRepository.countByEventId(event.getId());
        return EventResponse.builder()
                .id(event.getId())
                .name(event.getName())
                .description(event.getDescription())
                .status(event.getStatus())
                .startTime(event.getStartTime())
                .endTime(event.getEndTime())
                .teamCount(teamCount)
                .createdAt(event.getCreatedAt())
                .updatedAt(event.getUpdatedAt())
                .build();
    }
}

File: backend/services/eventService.js
const eventRepository = require('../repositories/eventRepository');
const teamRepository = require('../repositories/teamRepository');
const auditService = require('./auditService');
const { ResourceNotFoundException } = require('../middleware/errorHandler');

class EventService {
  async createEvent(request) {
    this._validateTimestamps(request.startTime, request.endTime);

    const rawPasskey = (request.passkey && request.passkey.trim())
      ? request.passkey.trim()
      : this._generateRandomPasskey();

    const passkeyHash = "HASH_" + this._hashCode(rawPasskey).toString(16);

    const event = await eventRepository.create({
      name: (request.name || '').trim(),
      description: request.description || '',
      status: 'DRAFT',
      passkeyHash,
      startTime: request.startTime || null,
      endTime: request.endTime || null
    });

    await auditService.logEvent(
      'EVENT_CREATED',
      event,
      null,
      null,
      JSON.stringify({ name: event.name }),
      'ADMIN'
    );

    return await this._mapToResponse(event);
  }

  async getEventById(eventId) {
    const event = await this.findEventOrThrow(eventId);
    return await this._mapToResponse(event);
  }

  async getAllEvents() {
    const events = await eventRepository.findAll();
    const responses = [];
    for (const ev of events) {
      responses.push(await this._mapToResponse(ev));
    }
    return responses;
  }

  async updateEvent(eventId, request) {
    this._validateTimestamps(request.startTime, request.endTime);

    const event = await this.findEventOrThrow(eventId);
    const updated = await eventRepository.update(eventId, {
      name: (request.name || event.name).trim(),
      description: request.description !== undefined ? request.description : event.description,
      status: request.status || event.status,
      startTime: request.startTime !== undefined ? request.startTime : event.startTime,
      endTime: request.endTime !== undefined ? request.endTime : event.endTime
    });

    await auditService.logEvent(
      'EVENT_UPDATED',
      updated,
      null,
      null,
      JSON.stringify({ status: updated.status }),
      'ADMIN'
    );

    return await this._mapToResponse(updated);
  }

  async updateEventStatus(eventId, status) {
    const event = await this.findEventOrThrow(eventId);
    const updated = await eventRepository.updateStatus(eventId, status);

    await auditService.logEvent(
      'EVENT_STATUS_CHANGED',
      updated,
      null,
      null,
      JSON.stringify({ newStatus: status }),
      'ADMIN'
    );

    return await this._mapToResponse(updated);
  }

  async findEventOrThrow(eventId) {
    const event = await eventRepository.findById(eventId);
    if (!event) {
      throw new ResourceNotFoundException(`Event not found with ID: ${eventId}`);
    }
    return event;
  }

  _validateTimestamps(startTime, endTime) {
    if (startTime && endTime && new Date(endTime) < new Date(startTime)) {
      const err = new Error('Event end time cannot be before start time');
      err.status = 400;
      throw err;
    }
  }

  _generateRandomPasskey() {
    return String(100000 + Math.floor(Math.random() * 900000));
  }

  _hashCode(str) {
    let hash = 0;
    for (let i = 0; i < str.length; i++) {
      const char = str.charCodeAt(i);
      hash = ((hash << 5) - hash) + char;
      hash |= 0;
    }
    return hash;
  }

  async _mapToResponse(event) {
    const teamCount = await teamRepository.countByEventId(event.id);
    return {
      id: event.id,
      name: event.name,
      description: event.description,
      status: event.status,
      startTime: event.startTime,
      endTime: event.endTime,
      teamCount,
      createdAt: event.createdAt,
      updatedAt: event.updatedAt
    };
  }
}

module.exports = new EventService();
